caching/base.py

```diff
@@ -104,28 +112,47 @@ def __iter__(self):
                 yield obj
             return
 
-        # Do the database query, cache it once we have all the objects.
-        iterator = self.iter_function()
+        # Use the special FETCH_BY_ID iterator if configured.
+        if config.FETCH_BY_ID and hasattr(self.queryset, 'fetch_by_id'):
+            iterator = self.queryset.fetch_by_id
 
+        # No cached results. Do the database query, and cache it once we have
+        # all the objects.
         to_cache = []
-        try:
-            while True:
-                obj = next(iterator)
-                obj.from_cache = False
-                to_cache.append(obj)
-                yield obj
-        except StopIteration:
-            if to_cache or config.CACHE_EMPTY_QUERYSETS:
-                self.cache_objects(to_cache, query_key)
-            raise
```

Comment:
This change loses the `StopIteration` re-raised here -- is that intentional?

Reply:
Ahh... I think I missed that (I thought I was just simplifying an iteration through a iterator). I'll have to see if I can find a way to test that, or maybe just revert back to the old version of the code.

Reply:
Ah, perhaps it doesn't matter? https://stackoverflow.com/questions/4019971/how-to-implement-iter-self-for-a-container-object-python#comment67775147_4019987

Reply:
Actually, this might be a part of Python that I just don't understand well. I'd think that I could just fix this by adding a `raise StopIteration` on line 128 (at the same indentation level as the `if` on line 126). But that seems superfluous, since it's the last line of the method and the iteration will end anyways. Am I missing something?

Reply:
https://www.python.org/dev/peps/pep-0479/#rationale